docs/releases.md

```diff
@@ -1,39 +1,157 @@
-# go-ipfs releases
+# `go-ipfs` Release Flow
 
-## Release Schedule
-go-ipfs is on a six week release schedule. Following a release, there will be
-five weeks for code of any type (features, bugfixes, etc) to be added. After
-the five weeks is up, a release candidate is tagged and only important bugfixes
-will be allowed up to release day.
+## Table of Contents
 
-## Release Candidate Checklist
-- [ ] CHANGELOG.md has been updated
-  - use `./bin/mkreleaselog` to generate a nice starter list
+- [Release Philosophy](#release-philosophy)
+- [Release Flow](#release-flow)
+- [Performing a Release](#performing-a-release)
+- [Release Version Numbers (aka semver)](#release-version-numbers-aka-semver)
+
+## Release Philosophy
+
+`go-ipfs` aims to have release every six weeks, two releases per quarter. During these 6 week releases, we go through 4 different stages of Release Candidates (RC) that gives us the opportunity to test the new version against our test envinronments, production envinronments, IPFS apps (e.g. Desktop and WebUI) and with our close partners that have IPFS running in Production, by leveraging their own test infrastructure.
+
+We might expand the six week release schedule in case of:
+- No new updates to be added
+- In case of a large community event that takes the core team availability away (e.g. IPFS Conf, Dev Meetings, IPFS Camp, etc)
+
+## Release Flow
+
+`go-ipfs` releases come in 4 stages:
+
+- **Release Stage 1 - Internal testing** - Test the Release against our testing infrastructure, including interoperability, integration, test lab, multiple runtimes and the apps we've built (WebUI, Desktop, NPM on IPFS, HTTP Client Libraries). The intent is to make this stage fully automated (and somewhat is already), until then, we manually check a list and ensure that all tests have been run
+- **Release Stage 2 - Invite IPFS partners to test** - Reach out to our partners (i.e. projects that have volunteered to support `go-ipfs` by using their own test infrastructure and tell us the results)
+- **Release Stage 3 - Announce to the broader community** - Communicate to the community that a new Release Candidate is ready and that everyone is welcome to test it with us
```

Comment:
Any real reason not to combine stages 2 and 3? Really, we usually ask the broader community (stage 3) first because they tend not to be using IPFS for mission critical operations. That is:

1. Advanced/eager users tend to find "shallow" bugs: api breakage, bad documentation, etc.
2. Partners tend to find deep bugs: performance regressions, rare crashes, etc.

Reply:
My rationale for this separation was based on my assumption that we are very limited in capacity to deal with incoming bug reports.

I expect that our early testers are more knowledgeable and therefore, more capable of debugging themselves part of the problem or at least providing more useful information, which will save us a ton of time and cover a ton of bugs. 

In contrast, I expect that the broad community to be able to also pick on bugs but provide less useful information. By doing early testers first, we save some time from debugging less documented/debugged bugs (which would overlap with the ones that our early testers will report).

Reply:
That said, I'm not holding any strong feelings towards the flow. Happy to take in the wisdom of the maintainers and iterate on it :)

Reply:
What about:

* Stage 2 - Announce a preliminary RC on IRC and invite the community to run sanity checks.
* Stage 3 - Pull in partners.
* Stage 4 - Announce the RC on twitter etc? I wonder if there's some way to allow, e.g., ipfs-desktop users to opt into this release.

I guess I mostly just want a chance to test it with _lower_ stakes than people relying on IPFS in production systems.

> My rationale for this separation was based on my assumption that we are very limited in capacity to deal with incoming bug reports.

Most of the bug reports we get from announcements on IRC are actually pretty good. They also tend to be pretty straight forward to diagnose as they aren't subtle performance issues.

Reply:
> lower stakes than people relying on IPFS in production systems.

We should not advise folks to run RC in production, rather just invite them to test on their test infra and QA if they have it built.

> Most of the bug reports we get from announcements on IRC are actually pretty good. They also tend to be pretty straight forward to diagnose as they aren't subtle performance issues.

This is good info to know. If you feel confident on merging Stage 2 and 3, then let's do it. Just confirm and I'll update it

Reply:
> We should not advise folks to run RC in production, rather just invite them to test on their test infra and QA if they have it built.

We _need_ to run our final RC in production before releasing it. Basically, we have two options:

1. We cut a release with a performance regression an wait a release cycle (or go back through the release cycle for a patch release).
2. We cut release _candidate_ and ask partners to start running it. This way, we can fix any performance regressions without starting a new release cycle.

This RC should be a true RC, a candidate for release.

I believe this comes down to an issue of naming. Really, the stages are:

1. Alpha - Internal: Small sample size, light and heavy workloads.
2. Beta - IRC: Medium sample size, light workloads.
3. RC1 - Partners: Medium sample size, heavy workloads.
4. RC2 - End users (twitter): Large sample size.

However, IPFS itself is "alpha" so this would be even more confusing.

I'd _love_ to get partners on board at the "beta" stage but I'm not sure how helpful that will be unless they're willing to build out simulations for us. Having more testing on light workloads won't help all that much.

Reply:
IMHO, RC1 (using @Stebalien terminology ^), should be run in production by more than just us. It doesn’t need to be nor should it be a full scale deployment of the RC1 version but to assume that deploying an RC1 to partner test/QA infra where the load is no where near what it is in production is going to lead us to believe things are fine when they are not. The burden that we place on our partners in this process should be as light as possible, and bringing them in at the RC1 stage is about ensuring that any changes made to go-ipfs do not lead to regressions in performance or functionality when running in their particular usecase/environment, which to me indicates that they should be able to deploy at least one instance of the RC1 to production without any dire consequences.

Reply:
It's sounds weird that certain releases are meant for partners. It sounds like giving early access to features even though it means "try it out". It should be very clear that anyone can try it out. I can also imagine that different partners (and users) may choose to test ipfs pre-releasees at different stages depending on circumstances.

Note that it is also weird that we give some partners a half-baked release to test. Either they test it by deploying it to their users, in which case they should not be given a half-baked release that can break their apps (even if they volunteer), or they have they're own testing environments, in which case they could directly master or a release branch continuously without waiting for a specific release stage.

Otherwise said:

  * Sounds good to have alpha/beta/rcs (we should have a release branch that includes all features in the release and only updates with bugfixes). We should be able to skip certain checkpoints sometimes (like hotfixes to specific issues).
  * Document what the level of maturity is after each checkpoint, so that everyone can test and try it out (partner or not).
  * Create a known and consistent venue to gather feedback on each release (i.e. an issue) and include information from our own testing in it.
  * Let people choose when they test based on their availability, feedback on the current state of the release and issues with the previous release.

Reply:
**We absolutely must not encourage partners to run a RC in production that can potentially harm their service and hurt their users.** What we hope is for partners to use their own Beta channels, Testing Infrastructure and potentially QA setup to let us know if the new release passes the tests.

Reply:
> We absolutely must not encourage partners to run a RC in production that can potentially harm their service and hurt their users.

Any release we make can potentially break something. This is an inviolate rule of software and we must not lie to ourselves about this. The goal here is to minimize the impact by deploying in stages, not to avoid all harm.

We definitely shouldn't ask our partners to run RCs that haven't been tested but we should absolutely ask them to run an RC that we believe to be "ready for release", to reduce the impact of any bugs they may encounter. That way they can roll back, we can fix the issue, and then move on with the release cycle without waiting for another release.

> What we hope is for partners to use their own Beta channels, Testing Infrastructure and potentially QA setup to let us know if the new release passes the tests.

This would be awesome but it's asking a lot from our partners and isn't currently the weakest part of our release process.

Reply:
You know, we probably _can_ just call these alpha, beta, etc. That should make this easier to explain to users:

* Dev - Bleeding Edge.
* Alpha - Feature Freeze.
* Beta - Tested but may still be buggy.
* RC1 - Release quality, small rollout.
* RC2 - Release quality, larger rollout.
* Release - Full rollout.

Reply:
See: https://github.com/ipfs/go-ipfs/pull/6496